client/src/main/scala/io/chrisdavenport/circuit/http4s/client/CircuitedClient.scala

```diff
@@ -1,13 +1,138 @@
 package io.chrisdavenport.circuit.http4s.client
 
 import io.chrisdavenport.circuit.CircuitBreaker
+import cats.syntax.all._
 import cats.effect._
+import cats.effect.concurrent._
 import org.http4s._
 import org.http4s.client._
-
+import cats._
+import scala.concurrent.duration._
 object CircuitedClient {
 
   def apply[F[_]: Bracket[*[_], Throwable]](cr: CircuitBreaker[F])(c: Client[F]): Client[F] = 
-    Client{req: Request[F] => Resource(cr.protect(c.run(req).allocated))}
+    generic(Function.const[CircuitBreaker[F], Request[F]](cr))(c)
+
+
+  def byKey[F[_]](
+    maxFailures: Int,
+    resetTimeout: FiniteDuration,
+    exponentialBackoffFactor: Double = 1,
+    maxResetTimeout: Duration = Duration.Inf,
+    modifications: CircuitBreaker[F] => CircuitBreaker[F] = {x: CircuitBreaker[F] => x},
+    shouldFail: (Request[F], Response[F]) => Boolean = defaultShouldFail[F](_, _)
+  )(client: Client[F])(implicit F: Sync[F], C: Clock[F]): F[Client[F]] = {
+    io.chrisdavenport.mapref.MapRef.ofSingleImmutableMap[F, RequestKey, CircuitBreaker.State](Map.empty[RequestKey, CircuitBreaker.State]).map{
+      mapref => 
+        val f : Request[F] => CircuitBreaker[F] = {req: Request[F] => 
+          val key = RequestKey.fromRequest(req)
+          val optRef = mapref(key)
+          val ref = new LiftedRefDefaultStorage(optRef, CircuitBreaker.Closed(0))
+          val cbInit = CircuitBreaker.unsafe(ref, maxFailures, resetTimeout, exponentialBackoffFactor, maxResetTimeout, F.unit, F.unit, F.unit, F.unit)
+          modifications(cbInit)
+        }
+        generic(f, shouldFail)(client)
+    }
+  }
+
+  def generic[F[_], A](
+    cbf: Request[F] => CircuitBreaker[F],
+    shouldFail: (Request[F], Response[F]) => Boolean = defaultShouldFail[F](_, _)
+  )(client: Client[F])(implicit F: Bracket[F, Throwable]): Client[F] = {
+    Client[F]{req: Request[F] => 
+      val cb = cbf(req)
+      Resource(
+        cb.protect(
+          client.run(req).allocated.flatMap{
+            case (resp, shutdown) => 
+              if (shouldFail(req, resp)) F.raiseError[(Response[F], F[Unit])](CircuitedClientThrowable(resp, shutdown))
+              else (resp, shutdown).pure[F]
+          }
+        ).handleErrorWith{
+          // The F here is not checked.
+          case e: CircuitedClientThrowable[F] @unchecked => (e.resp, e.shutdown).pure[F]
+          case e => F.raiseError(e)
+        }
+      )
+      
+    }
+  }
+
+  def defaultShouldFail[F[_]](req: Request[F], resp: Response[F]): Boolean = {
+    val _ = req
+    if (resp.status.code >= 500) true
```

Comment:
I think this is the safest default. Only if the servers response code says its responsible for the failure.

Reply:
`Status.ServerError(resp)` or `resp.status.responseClass === Status.ServerError` do this.